pythonPackages.pyscf: init at 1.7.1

drewrisinger opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Add dependency for upcoming qiskit-terra package #78772 @jonringer 
Depends on #78871 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

drewrisinger committed:
pythonPackages.pyscf: init at 1.7.0

Python-based Simulations of Chemistry Framework.

Requirement for qiskit-aqua.

drewrisinger commented:
@GrahamcOfBorg build python3Packages.pyscf

drewrisinger commented:
build fails, depending on PR #78871 (libcint)

timokau commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,67 @@
+    scipy
+  ];
+
+  PYSCF_INC_DIR = lib.string.makeLibraryPath [ libcint libxc ];
```
`nix-review pr 78872` shows:
```
error: while querying the derivation named 'python2.7-pyscf-1.7.0':
while evaluating the attribute 'out.outPath' at /home/timo/.cache/nixpkgs-review/pr-78872/nixpkgs/lib/customisation.nix:151:13:
while evaluating the attribute 'PYSCF_INC_DIR' of the derivation 'python2.7-pyscf-1.7.0' at /home/timo/.cache/nixpkgs-review/pr-78872/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:105:3:
attribute 'string' missing, at /home/timo/.cache/nixpkgs-review/pr-78872/nixpkgs/pkgs/development/python-modules/pyscf/default.nix:47:19
```

timokau commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+  # must download from GitHub to get the Cmake & C source files
+  src = fetchFromGitHub {
+    owner = pname;
```
This is a case of *incidental* equality. While `owner` and `repo` happen to be the same as `pname`, they do not *change* together. The  name might change because of a rename while staying in the same repo. Or we might decide to switch to a fork.

Therefore it doesn't make any sense to use `pname` for `owner` and `repo`. It doesn't hurt much either, but it is a bit of a pet-peeve of mine.

timokau commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,67 @@
+  pythonImportsCheck = [ "pyscf" ];
+  checkInputs = [ pytest ];
+  checkPhase = ''
+    pytest ./pyscf/tools/test
```
Is that the official way to run the tests? Doesn't a plain `pytest` suffice?

drewrisinger replied:
Sort of? I just looked into this, and it appears to have a lot of individual unittests, but my pytest always hangs on collecting the tests for some reason. Not able to easily trace where it's hanging.

It also has an extensive suite of examples, but most seem to be unguarded scripts (no if __name__ == "__main__"), and not in a standard test-runner format.

Will update when I find out more.

drewrisinger replied:
I ran the tests in Travis CI build. I hadn't realized tests were distributed through the source dirs.
https://github.com/pyscf/pyscf/blob/master/.travis.yml

timokau reviewed (changes requested):
Looks pretty good to me, but there is one serious issue (wrong name for the library function) and one minor nitpick.

drewrisinger committed:
pyscf: tests running

drewrisinger committed:
pyscf: cleanup unused code

drewrisinger committed:
pyscf: break check into multiple hooks

drewrisinger committed:
xcfun: init TODO

drewrisinger committed:
xcfun: init at 1.0.0

Exchange-correlation functionals with arbitrary order derivatives.

Library dependency of pythonPackages.pyscf.

drewrisinger committed:
xcfun: build library as .so vs .a

drewrisinger committed:
python.xcfun: remove

drewrisinger committed:
WIP: fix xcfun/pyscf

drewrisinger committed:
pythonPackages.pyscf: 1.6.6 -> 1.7.1

drewrisinger committed:
libcint: add pythonPackages.pyscf build flags

As far as I know, libcint is only used for PySCF, so this just corrects
the build flags that it should have to build properly for that.

drewrisinger committed:
pythonPackages.pyscf: make most tests work. few errors remain

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
+  prePatch = ''
+    mkdir -p ./pyscf/lib/deps/include ./pyscf/lib/deps/lib
+    ln -s ${lib.strings.makeSearchPath "include" [xcfun] }/xc.h ./pyscf/lib/deps/include/xcfun.h
+    ln -s ${lib.strings.makeLibraryPath [ xcfun ] }/libxc.so ./pyscf/lib/deps/lib/libxcfun.so
```
I think you mean
```suggestion
    ln -s ${lib.getLib xcfun}/libxc.so ./pyscf/lib/deps/lib/libxcfun.so
```

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
+
+  prePatch = ''
+    mkdir -p ./pyscf/lib/deps/include ./pyscf/lib/deps/lib
+    ln -s ${lib.strings.makeSearchPath "include" [xcfun] }/xc.h ./pyscf/lib/deps/include/xcfun.h
```
```suggestion
    ln -s ${lib.getDev xcfun}/include/xc.h ./pyscf/lib/deps/include/xcfun.h
```

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
+  # Configure CMake to build C files in pyscf/lib
+  preConfigure = ''
+    pushd pyscf/lib
+  '';
```
might not work depending on PWD
```suggestion
  cmakeDir = "pyscf/lib";
```

drewrisinger replied:
I tried this, but the rest of the python build process expects files in ``PWD/pyscf/lib/build``.

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
   ];
 
-  PYSCF_INC_DIR = lib.strings.makeLibraryPath [ libcint libxc ];
+  PYSCF_INC_DIR = lib.strings.makeLibraryPath [ libcint libxc ] + ":" + lib.strings.makeSearchPath "include" [ xcfun ];
```
```suggestion
  PYSCF_INC_DIR = lib.makeSearchPath "include" (map lib.getDev [
    libcint
    libxc
    xcfun
  ];
```

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
+  # Build C dependencies
+  preBuild = ''
+    pushd pyscf/lib/build
+    export LD_LIBRARY_PATH=$(pwd)/lib/deps/include:$LD_LIBRARY_PATH
```
I'm confused, LD_LIBRARY_PATH is usually for libraries,  and include dirs are usually for headers....

drewrisinger replied:
Yeah, I thought I needed it b/c pyscf was looking for a library on the path, and I thought it was getting put in that dir. builds fine without that export.

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,57 +1,143 @@
 buildPythonPackage rec {
   pname = "pyscf";
-  version = "1.7.0";
+  version = "1.6.6";
```
according to the diff, we are doing `1.7.0` -> `1.6.6`, which doesnt fit the title

drewrisinger replied:
Yes, that was a mistake. Was testing if lowering version helped.

drewrisinger commented:
I'm not sure if this should be included in the main Nix repo. Right now the only reason for this package's existence is a plugin to ``pythonPackages.qiskit-aqua``, but I've fixed that package so this isn't a strict requirement. Maybe Nix User Repository would be a better fit? I don't have strong feelings.

drewrisinger committed:
pythonPackages.pyscf: cleanup with code review suggestions

drewrisinger commented:
@GrahamcOfBorg eval

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,164 @@
+    ln -s ${lib.getLib xcfun}/lib/libxc.so ./pyscf/lib/deps/lib/libxcfun.so
+    substituteInPlace pyscf/rt/__init__.py --replace "from tdscf import *" "from pyscf.tdscf import *"
+  '';
+
```
```suggestion
```

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,164 @@
+    pushd pyscf/lib/build
+    make
+    popd
+  '';
```
this should work as well
```suggestion
  makeFlags = [ "-C pyscf/lib/build" ];
```

jonringer commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -0,0 +1,164 @@
+  # Tests take about 30 mins to run
+  preCheck = ''
+    # HACK: Move compiled libraries to test dir so pyscf import mechanism can find them
+    cp ./dist/tmpbuild/pyscf/pyscf/lib/*.so ./pyscf/lib/
```
this might not be needed anymore, the `pythonImportsCheck` hook was patched so it would change PWD to $out

drewrisinger commented:
@jonringer thoughts on above question? https://github.com/NixOS/nixpkgs/pull/78872#issuecomment-606676366

jonringer commented:
> I'm not sure if this should be included in the main Nix repo. Right now the only reason for this package's existence is a plugin to `pythonPackages.qiskit-aqua`, but I've fixed that package so this isn't a strict requirement. Maybe Nix User Repository would be a better fit? I don't have strong feelings.

I think moving to the NUR is great for now. There is a lot going on with these package additions, and I'm not sure if either of us want to get it to that level of a "polished" shine.

drewrisinger commented:
Agreed, thanks. It's always felt like an ugly, difficult package to me, so I don't really mind removing it. Related: Would it be worth removing ``libcint`` (#78771), as this is the only thing that required it?

jonringer commented:
> Agreed, thanks. It's always felt like an ugly, difficult package to me, so I don't really mind removing it. Related: Would it be worth removing `libcint` (#78771), as this is the only thing that required it?

If burden of maintainership is low (I'm not super familiar with the oCaml nix ecosystem), then I would say leave it.

Python packages are a pretty large maintainer burden due to how fluid most breaking changes are.

drewrisinger commented:
Ack @jonringer. Wrong PR, I meant ``libcint`` (C library) #78871, whoops.

jonringer commented:
Statement still applies, doesn't look like an extream burden. I would just leave

drewrisinger commented:
Future reference: moved to https://github.com/drewrisinger/nur-packages.
https://github.com/drewrisinger/nur-packages/commit/498953542dd1a5e6032fbd3d6713e846711c2c98
